### PR TITLE
interfaces: compose the base declaration from interfaces

### PR DESCRIPTION
**Pull request by zyga**
Whenever a new interface is created the developer has to edit a number
of files. With the recent improvements those have shrinked to, almost,
just the interface module. This patch reduces this to *exactly* just the
new interface module. Now each interface can augment the base
declaration straight from the module file.

The split change nothing in how the policy looks like but sets the stage
for composing it out of meta-data coming from individual interfaces.

Upcoming branch will move the base declaration into separate interface files.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: interfaces/core.go [hunk 1]
```diff
@@ -142,6 +142,9 @@ type MetaData struct {
 	ImplicitOnCore bool `json:"implicit-on-core,omitempty"`
 	// ImplicitOnClassic controls if a slot is automatically added to classic systems.
 	ImplicitOnClassic bool `json:"implicit-on-classic,omitempty"`
+
+	BaseDeclarationPlugs string
```
Comment (morphis) [inline interfaces/core.go]: A description of both variables would be nice

File: interfaces/policy/basedeclaration.go [hunk 2]
```diff
@@ -658,9 +669,45 @@ slots:
         - core
 `
 
+func ComposeBaseDeclaration(ifaces []interfaces.Interface) ([]byte, error) {
+	var buf bytes.Buffer
+	// Trim newlines at the end of the string. All the elements may have
+	// spurious trailing newlines. All elements start with a leading newline.
+	// We don't want any blanks as that would no longer parse.
+	tr := func(s string) string { return strings.TrimRight(s, "\n") }
```
Comment (morphis) [inline interfaces/policy/basedeclaration.go]: Would like a better name for `tr` here as it doesn't say much if you look at something like `tr(baseDeclarationHeader)`. Something like `trimTrailingNewLine` or so would make it obvious fo the reader.

File: interfaces/policy/basedeclaration_test.go [hunk 3]
```diff
@@ -760,3 +761,14 @@ plugs:
 	err = cand.Check()
 	c.Check(err, NotNil)
 }
+
+func (s *baseDeclSuite) TestComposeBaseDeclaration(c *C) {
+	decl, err := policy.ComposeBaseDeclaration(nil)
```
Comment (morphis) [inline interfaces/policy/basedeclaration_test.go]: Are the other tests exercising ComposeBaseDeclaration too?

Comment (zyga) [inline interfaces/policy/basedeclaration_test.go]: Yes, they do

File: interfaces/core.go (hunk 1 above)
Comment (zyga) [inline interfaces/core.go]: Sure, I'll add those in a moment

File: interfaces/policy/basedeclaration.go (hunk 2 above)
Comment (zyga) [inline interfaces/policy/basedeclaration.go]: Done

File: interfaces/policy/basedeclaration.go [hunk 4]
```diff
@@ -658,9 +669,48 @@ slots:
         - core
 `
 
+func trimTrailingNewline(s string) string {
+	return strings.TrimRight(s, "\n")
+}
+
+func ComposeBaseDeclaration(ifaces []interfaces.Interface) ([]byte, error) {
```
Comment (pedronis) [inline interfaces/policy/basedeclaration.go]: why public and without doc? it could be private and then tested using export_test?

Comment (zyga) [inline interfaces/policy/basedeclaration.go]: Ah, indeed. I don't need this to be public actually. I'll make it private

Comment (zyga) [inline interfaces/policy/basedeclaration.go]: Done now

**Comment by zyga** [comment]
@jdstrand I have that in another branch, I didn't want to make the review too hard. With https://github.com/zyga/snapd/tree/feature/metadata-defines-base-policy it will be exactly as you had hoped.

**Comment by pedronis** [comment]
@zyga that's ok there's another branch that refines this but then the description of this one that will go in the commit is very confusing

**Pull request merged**